### DISCUSSION TITLE
Check for existence of an id

### DISCUSSION THREAD

**Discussion by vishal-kvn** [milvus-io/milvus#7183] [Q&A and General discussion] [answered]
👋 I am working on adding a uniqueness check ie prevent vector insertion if the corresponding id already exists in the database. This prevents duplicates from being returned at query time. In order to do this, I am following these steps -

- Check for existing vectors by using `get_entity_by_id`
- Insert only vectors whose ids are not present in the previous step

This works but is very expensive for large databases(>100M) and also there is a limitation around the number of ids that can be used at a time with `get_entity_by_id`.

- Milvus database version: 1.1.1
- partition_tag: default

Based on the above description what is the recommended way to check for existence of an id in the Milvus?

**Comment by yhmo** [comment, marked as answer]
The get_entity_by_id() only support input 1000 ids on time. And it need to read raw data from raw file. Its performance is not good. But currently there is only one way to check id existence, just like the steps you are following.

A better solution is: provide a new api named `check_ids_exist()`, user input id array, the api return a boolean array, so that the client can know which id is exist or not. This api no need to read raw data, it only check the bloom filter and segment row id array, all this work can be done in memory.

But now we only fix critical bug for v1.x, no new feature will be implemented.

**Comment by hadim** [comment]
I would also be interested in that feature but for Milvus 2. @yhmo is that already possible to perform `check_ids_exist()` using the Milvus bloom filter?

**Reply by yhmo** [reply]
In 2.0, there is an api query(), can fetch entity data, or check entity existence.
The api definition:
`def query(self, expr, output_fields=None, partition_names=None, timeout=None):`

The "expr" is an string expression to filter entities, the "output_fields" is to return entity data such as vectors or scalar field.

Pymilvus example:
To check entity existence, just let the "output_fields" = None:
```
collection = Collection(name)
collection.query(expr = "id in [1, 2, 3, 4]", output_fields = None)
```

To return vector data, specify vector field name in the "output_fields":
```
collection = Collection(name)
collection.query(expr = "id in [1, 2, 3, 4]", output_fields = ["vec_field"])
```

**Reply by yhmo** [reply]
Currently, the bloom filter is not yet available for 2.0. The query() filter the id by brute-force search. We are working on the "delete" feature, the bloom filter will be integrated by this feature.

**Reply by hadim** [reply]
Thanks.

Will the bloom filter be accessible via the Python SDK? The query brute force approach seems not scalable to billions of rows.

**Reply by xiaofan-luan** [reply]
> Thanks.
> 
> Will the bloom filter be accessible via the Python SDK? The query brute force approach seems not scalable to billions of rows.

Uhh, since bf has false positive, this may not be the best solution for check id exist?

**Reply by xiaofan-luan** [reply]
We are thinking about introducing another keyvalue store, for instance foundation DB or TiKV for maintain rows in milvus, how do you think about this?

**Reply by hadim** [reply]
I have multiple usecases for such functions. Some of them can live with a small false positive rate if the speed is very high? Others need exact query, in that case a key value store seems reasonable.

I saw you're going to use Arrow. Is it what you are talking for the KV store? Have you already performed preliminary benchmark to assess scalabiliity?

**Reply by hadim** [reply]
(if the bloom filter is cheap for you to bring to the SDKs, that would be nice!)

**Reply by xiaofan-luan** [reply]
we've already added logic to generate bloom filters while datanode create files, currently bloomfilter is only used for delete operation. 
We should support two index types 1) Bloomfilter index 2)Kv Index for checking id  existing. How does that sound?

**Reply by xiaofan-luan** [reply]
Feel free to investigate and contribute if you are interested. Many users are looking forward to primary key deduplication
